j5/backends/hardware/sb/arduino.py

```diff
@@ -0,0 +1,281 @@
+            raise ValueError(f"Pin {identifier} mode needs to be DIGITAL_OUTPUT"
+                             f"in order to set the digital state.")
+        self._digital_pins[identifier].state = state
+        self._update_digital_pin(identifier)
+
+    def get_gpio_pin_digital_state(self, identifier: int) -> bool:
+        """Get the last written state of the GPIO pin."""
+        if identifier >= FIRST_ANALOGUE_PIN:
+            raise NotSupportedByHardwareError(
+                "Digital functions not supported on analogue pins",
+            )
+        if self._digital_pins[identifier].mode is not GPIOPinMode.DIGITAL_OUTPUT:
+            raise ValueError(f"Pin {identifier} mode needs to be DIGITAL_OUTPUT"
+                             f"in order to read the digital state.")
+        return self._digital_pins[identifier].state
+
+    def read_gpio_pin_digital_state(self, identifier: int) -> bool:
+        """Read the digital state of the GPIO pin."""
+        if identifier >= FIRST_ANALOGUE_PIN:
+            raise NotSupportedByHardwareError(
+                "Digital functions not supported on analogue pins",
+            )
+        if self._digital_pins[identifier].mode not in (
+            GPIOPinMode.DIGITAL_INPUT,
+            GPIOPinMode.DIGITAL_INPUT_PULLUP,
+        ):
+            raise ValueError(f"Pin {identifier} mode needs to be DIGITAL_INPUT_*"
+                             f"in order to read the digital state.")
+        results = self._command("R", str(identifier))
+        if len(results) != 1:
+            raise CommunicationError(f"Invalid response from Arduino: {results}")
+        result = results[0]
+        if result == "H":
+            return True
+        elif result == "L":
+            return False
+        else:
+            raise CommunicationError(f"Invalid response from Arduino: {result}")
+
+    def read_gpio_pin_analogue_value(self, identifier: int) -> float:
+        """Read the analogue voltage of the GPIO pin."""
+        if identifier < FIRST_ANALOGUE_PIN:
+            raise NotSupportedByHardwareError(
+                "Analogue functions not supported on digital pins",
+            )
+        if identifier >= FIRST_ANALOGUE_PIN + 4:
+            raise NotSupportedByHardwareError(
+                f"Arduino Uno firmware only supports analogue pins 0-3 (IDs 14-17)",
+            )
+        analogue_pin_num = identifier - 14
+        results = self._command("A")
+        for result in results:
+            pin_name, reading = result.split(None, 1)
+            if pin_name == f"a{analogue_pin_num}":
+                voltage = (int(reading) / 1024.0) * 5.0
+                return voltage
+        raise CommunicationError(f"Invalid response from Arduino: {results}")
+
+    def write_gpio_pin_dac_value(self, identifier: int, scaled_value: float) -> None:
+        """Write a scaled analogue value to the DAC on the GPIO pin."""
+        # Uno doesn't have any of these.
+        raise NotImplementedError
```

Comment:
This should be a `HardwareNotSupported` error

Reply:
4d6756c14582b00cde4943e12e6ce98cde37d561